test/rosdep_repo_check/make_rule.py

```diff
@@ -0,0 +1,226 @@
+            distros.append(distro_name)
+    return tuple(distros)
+
+
+def get_index() -> rosdistro.index.Index:
+    return rosdistro.get_index(rosdistro.get_index_url())
+
+
+def release_platforms(
+        index: rosdistro.index.Index,
+        distro_names: Sequence[str]
+    ) -> Dict[str, Sequence[str]]:
+
+    platforms = {}
+    for distro in distro_names:
+        for distro_file in rosdistro.get_distribution_files(index, distro):
+            for os, releases in distro_file.release_platforms.items():
+                if os not in platforms:
+                    platforms[os] = releases
+                else:
+                    for release in releases:
+                        if release not in platforms[os]:
+                            platforms[os].append(release)
+    return platforms
+
+
+
+class Rule:
+
+    def __init__(self, *, key=None):
+        # ex: 'ubuntu': ['focal', 'jammy']
+        self._platforms: Dict[str, List[str]] = {}
+        # ex: ('ubuntu', 'focal'): 'cmake'
+        self._packages: Dict[Tuple[str, str], Optional[str]] = {}
+        # Override the key of the rosdep rule
+        self._key = key
+
+    def set_package_for_platform(
+            self, os: str, release: str, package_name: Optional[str] = None
+        ) -> None:
+        os = str(os)
+        release = str(release)
+        if package_name:
+            package_name = str(package_name)
+        if os not in self._platforms.keys():
+            self._platforms[os] = []
+        if release not in self._platforms[os]:
+            self._platforms[os].append(release)
+        self._packages[(os, release)] = package_name
+
+    def _calculate_key(self) -> Optional[str]:
+        # Prioritize name of ubuntu package as the key name
+        for os in ('ubuntu', 'debian'):
+            if os in self._platforms.keys():
+                for release in self._platforms[os]:
+                    package = self._packages[(os, release)]
+                    if package:
+                        return package
+
+    @property
+    def key(self) -> Optional[str]:
+        if not self._key:
+            self._key = self._calculate_key()
+        return self._key
+
+    def as_dict(self, minimal=True):
+        if self.key is None:
+            raise ValueError('Rule has no key')
+        yaml_dict = {}
+
+        # First pass, put all the data in
+        for os, releases in self._platforms.items():
+            if os not in yaml_dict:
+                yaml_dict[os] = {}
+            for release in releases:
+                yaml_dict[os][release] = [self._packages[(os, release)]]
+
+        # Second pass, minimize keys
+        if minimal:
+            for os, release_packages in yaml_dict.items():
+                package_names = set([l[0] for l in release_packages.values()])
+                if len(package_names) == 1:
+                    # All releases have the same package name
+                    yaml_dict[os] = [package_names.pop()]
```

Comment:
This section doesn't handle `null` rules very well, and results in a list containing `null` rather than simply `null`. Here's one way to address that.
```suggestion
                package = self._packages[(os, release)]
                yaml_dict[os][release] = [package] if package else None

        # Second pass, minimize keys
        if minimal:
            for os, release_packages in yaml_dict.items():
                package_names = set([l if l is None else l[0] for l in release_packages.values()])
                if len(package_names) == 1:
                    # All releases have the same package name
                    package = package_names.pop()
                    yaml_dict[os] = [package] if package else None
```